Raspberry Pi: Segmentation fault when running server

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [mailing list](https://groups.google.com/forum/#!forum/vault-tool), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).
-->


<!-- Uncomment this section if this is a feature request. Include or exclude other sections as deemed appropriate.
**Feature Request:**
-->

**Environment:**

* Vault Version: Vault v0.8.2 ('9afe7330e06e486ee326621624f2077d88bc9511')
* Operating System/Architecture: Raspberry Pi 2 and 3 with Raspbian Stretch, Linux 4.9.41-v7+ #1023 SMP Tue Aug 8 16:00:15 BST 2017 armv7l GNU/Linux

**Vault Config File:**
none, ./vault server -dev

**Startup Log Output:**
./vault server -dev
==> Vault server configuration:

                     Cgo: disabled
         Cluster Address: https://127.0.0.1:8201
              Listener 1: tcp (addr: "127.0.0.1:8200", cluster address: "127.0.0.1:8201", tls: "disabled")
               Log Level: info
                   Mlock: supported: true, enabled: false
        Redirect Address: http://127.0.0.1:8200
                 Storage: inmem
                 Version: Vault v0.8.2
             Version Sha: 9afe7330e06e486ee326621624f2077d88bc9511

panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x11540]

goroutine 37 [running]:
sync/atomic.storeUint64(0x1238017c, 0x0, 0x0)
	/goroot/src/sync/atomic/64bit_arm.go:20 +0x3c
github.com/hashicorp/vault/vault.(*ExpirationManager).Restore.func1(0x12380140, 0x12872fe8, 0x125c1500)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:279 +0x34
panic(0x10b8978, 0x1e0bee8)
	/goroot/src/runtime/panic.go:491 +0x204
sync/atomic.storeUint64(0x1238017c, 0x0, 0x0)
	/goroot/src/sync/atomic/64bit_arm.go:20 +0x3c
github.com/hashicorp/vault/vault.(*ExpirationManager).Restore(0x12380140, 0x125c1500, 0x0, 0x0, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:394 +0x4f0
created by github.com/hashicorp/vault/vault.(*Core).setupExpiration
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:122 +0x214

**Expected Behavior:**
running dev server

**Actual Behavior:**
segmentation fault

**Steps to Reproduce:**
download latest arm binary and start dev server: ./vault server -dev

**Important Factoids:**
<!-- Describe any atypical environment setup, if any. -->

**References:**
<!-- Link to any references, such as GitHub issues or pull requests. -->

--- a/vault/expiration.go
+++ b/vault/expiration.go
@@ -51,12 +51,6 @@ const (
 // If a secret is not renewed in timely manner, it may be expired, and
 // the ExpirationManager will handle doing automatic revocation.
 type ExpirationManager struct {
-	// A Go bug requires 64-bit atomic variables to be placed at the beginning
-	// of the struct. The first word is guaranteed to be 64-bit aligned by the
-	// compiler which is required for certain CPU architectures.
-	// https://godoc.org/sync/atomic#pkg-note-bug
-	restoreMode int64
-
 	router     *Router
 	idView     *BarrierView
 	tokenView  *BarrierView
@@ -66,9 +60,9 @@ type ExpirationManager struct {
 	pending     map[string]*time.Timer
 	pendingLock sync.RWMutex
 
-	tidyLock sync.Mutex
-	tidyMode int
+	tidyLock int32
 
+	restoreMode        int32
 	restoreModeLock    sync.RWMutex
 	restoreRequestLock sync.RWMutex
 	restoreLocks       []*locksutil.LockEntry
@@ -154,7 +148,7 @@ func (m *ExpirationManager) unlockLease(leaseID string) {
 
 // inRestoreMode returns if we are currently in restore mode
 func (m *ExpirationManager) inRestoreMode() bool {
-	return atomic.LoadInt64(&m.restoreMode) == 1
+	return atomic.LoadInt32(&m.restoreMode) == 1
 }
 
 // Tidy cleans up the dangling storage entries for leases. It scans the storage
@@ -170,21 +164,12 @@ func (m *ExpirationManager) Tidy() error {
 
 	var tidyErrors *multierror.Error
 
-	// Allow only one tidy operatioon at a time
-	m.tidyLock.Lock()
-	if m.tidyMode == 1 {
-		m.tidyLock.Unlock()
+	if !atomic.CompareAndSwapInt32(&m.tidyLock, 0, 1) {
 		m.logger.Warn("expiration: tidy operation on leases is already in progress")
 		return fmt.Errorf("tidy operation on leases is already in progress")
 	}
-	m.tidyMode = 1
-	m.tidyLock.Unlock()
 
-	defer func() {
-		m.tidyLock.Lock()
-		m.tidyMode = 0
-		m.tidyLock.Unlock()
-	}()
+	defer atomic.CompareAndSwapInt32(&m.tidyLock, 1, 0)
 
 	m.logger.Info("expiration: beginning tidy operation on leases")
 	defer m.logger.Info("expiration: finished tidy operation on leases")
@@ -289,7 +274,7 @@ func (m *ExpirationManager) Restore(errorFunc func(), loadDelay time.Duration) (
 		// if restore mode finished successfully, restore mode was already
 		// disabled with the lock. In an error state, this will allow the
 		// Stop() function to shut everything down.
-		atomic.StoreInt64(&m.restoreMode, 0)
+		atomic.StoreInt32(&m.restoreMode, 0)
 
 		switch {
 		case retErr == nil:
@@ -404,7 +389,7 @@ func (m *ExpirationManager) Restore(errorFunc func(), loadDelay time.Duration) (
 	m.restoreModeLock.Lock()
 	m.restoreLoaded = sync.Map{}
 	m.restoreLocks = nil
-	atomic.StoreInt64(&m.restoreMode, 0)
+	atomic.StoreInt32(&m.restoreMode, 0)
 	m.restoreModeLock.Unlock()
 
 	m.logger.Info("expiration: lease restore complete")

--- a/vault/expiration.go
+++ b/vault/expiration.go
@@ -51,6 +51,12 @@ const (
 // If a secret is not renewed in timely manner, it may be expired, and
 // the ExpirationManager will handle doing automatic revocation.
 type ExpirationManager struct {
+	// A Go bug requires 64-bit atomic variables to be placed at the beginning
+	// of the struct. The first word is guaranteed to be 64-bit aligned by the
+	// compiler
+	// https://godoc.org/sync/atomic#pkg-note-bug
+	restoreMode int64
+
 	router     *Router
 	idView     *BarrierView
 	tokenView  *BarrierView
@@ -60,10 +66,9 @@ type ExpirationManager struct {
 	pending     map[string]*time.Timer
 	pendingLock sync.RWMutex
 
-	tidyLock int64
+	tidyLock sync.Mutex
+	tidyMode int
 
-	// A set of locks to handle restoration
-	restoreMode        int64
 	restoreModeLock    sync.RWMutex
 	restoreRequestLock sync.RWMutex
 	restoreLocks       []*locksutil.LockEntry
@@ -166,12 +171,21 @@ func (m *ExpirationManager) Tidy() error {
 
 	var tidyErrors *multierror.Error
 
-	if !atomic.CompareAndSwapInt64(&m.tidyLock, 0, 1) {
+	// Allow only one tidy operatioon at a time
+	m.tidyLock.Lock()
+	if m.tidyMode == 1 {
+		m.tidyLock.Unlock()
 		m.logger.Warn("expiration: tidy operation on leases is already in progress")
 		return fmt.Errorf("tidy operation on leases is already in progress")
 	}
+	m.tidyMode = 1
+	m.tidyLock.Unlock()
 
-	defer atomic.CompareAndSwapInt64(&m.tidyLock, 1, 0)
+	defer func() {
+		m.tidyLock.Lock()
+		m.tidyMode = 0
+		m.tidyLock.Unlock()
+	}()
 
 	m.logger.Info("expiration: beginning tidy operation on leases")
 	defer m.logger.Info("expiration: finished tidy operation on leases")

--- a/vault/expiration.go
+++ b/vault/expiration.go
@@ -81,7 +81,6 @@ type ExpirationManager struct {
 func NewExpirationManager(router *Router, view *BarrierView, ts *TokenStore, logger log.Logger) *ExpirationManager {
 	if logger == nil {
 		logger = log.New("expiration_manager")
-
 	}
 
 	exp := &ExpirationManager{

--- a/vault/expiration.go
+++ b/vault/expiration.go
@@ -53,7 +53,7 @@ const (
 type ExpirationManager struct {
 	// A Go bug requires 64-bit atomic variables to be placed at the beginning
 	// of the struct. The first word is guaranteed to be 64-bit aligned by the
-	// compiler
+	// compiler which is required for certain CPU architectures.
 	// https://godoc.org/sync/atomic#pkg-note-bug
 	restoreMode int64
 

